### ISSUE TITLE
RFC Consider just using reStructured Text for hand written docs

### ISSUE THREAD

**Issue by weefuzzy** [flucoma/flucoma-docs#29]
At the moment we have a slightly bizarre system where the hand-written portion of the documentation is a YAML file, certain portions of which can contain rst markup to enable inline formatting in the generated reference for whatever host. Like all the best decisions, it came from successive needs to solve the problem in front of us quickly, but isn't exactly friendly. 

* Two formats in one file is confusing at the best of times 
* Both YAML and rst treat white space as semantically meaningful, but in different ways. This makes mistakes easy to produce, and not always easy to catch or diagnose. 

On the upside, YAML is a good match for the kind of structured data we want to present (i.e. about what controls and messages a Client exposes). The general use case for rst is more general, so specifying how we would want to write such structures in a way that is robust and checkable requires some thought. 

This change shouldn't have any effect on the overall documentation workflow, save for changing the format that humans have to work with. The generating stage should still work in the same way (although this *would* open the door for a future migration to sphinx if we wanted)

The obvious thing to do is follow what `docutils` and `sphinx` do for documenting code in various languages, and use the same sorts of rst entities to do the same sorts of work. Concretely, let's say we have some Client whose YAML currently looks like
```yaml
digest: Open a wormhole to the other side of the universe
see-also: Singularity 
description: Wormhole is a way of tearing the fabric of space-time
parameters:
  name:
    description: The name of the Wormhole. This is unique between all Wormhole.
messages:
  collapse: 
    description: collapse the wormhole, possibly with *devastating* effects
    args: 
      - name: quantum 
      - description: something quantum 
```

Man, that's worse than I remember 😆 I think the rst would look something like: 

```reStructuredText 
:digest: Open a wormhole to the other side of the universe
:see-also: Singularity 
:description: Wormhole is a way of tearing the fabric of space-time

.. parameter:: name

   The name of the Wormhole. This is unique between all Wormhole.

.. message:: quantum 

   collapse the wormhole, possibly with *devastating* effects

   :arg quantum: something quantum 

```

Which is no shorter but perhaps clearer? The slightly weird syntax
```reStructuredText
.. <something>::
```

is what's called a `directive` in rst, and seems to be the canonical way to represent some structured entity (see how the `message` directive above has a an inset `:arg <name>:` field`)   

(note that `parameter`s are about to be re-dubbed `control`s in a future Client API re-do, so an actual doc format would probably anticipate that).

@jamesb93 @tremblap @tedmoore any thoughts about the relative ickiness of these ways of spelling docs for Clients?

**Comment by jamesb93** [comment]
A singular format is better in the long run and YAML isn't _that_ readable once the data becomes as nested and dense as it is. For me, the consideration of readability goes in favour of RST anyway because the blocks are much more structured than YAML which is quite homogenous.

It feels a bit `eval()` to me, (read icky), when another dialect (RST) is interpreted by another program as a string somewhere else. Solidifying our efforts around a singular expressive format would be both easier to work with, validate, parse and otherwise reason about in terms of its role in the bigger pipeline of <big>DOCS</big>. It will also leave a more self-documenting code-base IMO too.

I am well for this.

**Comment by weefuzzy** [comment]
I should add that I think automatic conversion from our existing YAML into rst is trivial, so I'm not suggesting that humans need to rewrite all the existing documentation

**Comment by jamesb93** [comment]
That is even better.

**Comment by tremblap** [comment]
What I suffered most when writing the yaml was the fact my editor didn't support some sort of proper autotab, encapsulation, folding, etc so long documents start to be hard to parse in my little head. So I'm all for streamlining and making it more readable, but even better if it is a format know enough to Atom+friends to be able to format and fold.

**Comment by weefuzzy** [comment]
You need the [language-yaml](https://atom.io/packages/language-yaml) package installed on atom – the same would apply to [reStructured Text](https://atom.io/packages/language-restructuredtext). Atom doesn't know about very many things out of the box.

**Comment by tremblap** [comment]
I tried it and it didn't help much. Is reStructured Text better? What would be neat is to push here one (small) (handmade) example of a file so I could compare. I reckon you'll need to do one to get the translation script ready, or do you plan to attack the larger task from the outset?

**Comment by weefuzzy** [comment]
> I tried it and it didn't help much. 

It's got folding, and knows how to complain about certain sorts of indentation mistake, much like the JSON support.  

>  Is reStructured Text better? 

The Atom support is pretty much the same. It will do folding of directives and knows about whether some entities are properly formed or not. 

What we actually need (and fortunately what the current PR will enable) is a way of validating on the command line newly hand-written hand written files against the expectations implied by the JSON (which is the canonical source for being able to check that all controls, messages, args are documented), e.g 
```bash 
$   ./check_doc OpenWormhole
WARNING: 'digest' missing
WARNING: message 'collapse' is missing documentation 
```
sort of thig

>  What would be neat is to push here one (small) (handmade) example of a file so I could compare.

I would do an automated translator straight off the bat so that I could have tests to verify content parity with the originals as I go.  

The made example above was meant to be a starting point for comparison. I guess one of the minimal Clients like `Gain` could be changed with a couple of regex passes

**Comment by weefuzzy** [comment]
The current content for `Gain` could look like 
```reStructuredText
:digest: Real-Time Buffered Gain Changer
:sc-categories:  UGens>Algebraic, Libraries>FluidDecomposition, UGens>Buffer
:sc-related:  Guides/FluCoMa, Guides/FluidDecomposition,Classes/UnaryOpFunction
:see-also:  
:output:  Same as input, multiplied by the gain factor.
:sc-code: FluidGain.scd

This class implements a sanity test for the FluCoMa Real-Time Client Wrapper.

.. parameter::  in 

      The input to be processed

.. parameter:: gain

      Audio or control rate change of the gain.
```

**Comment by tremblap** [comment]
this looks very much more readable. What is the file suffix, so I can try it? I presume table support and such niceties exists too?

> knows about whether some entities are properly formed or not.

that never worked for me.

**Comment by weefuzzy** [comment]
> this looks very much more readable. What is the file suffix, so I can try it? 

`.rst` but you'll need to install the atom package to get any help from your editor 

> I presume table support and such niceties exists too?

This is indicative of the fact that the current situation is confusing. We already have table support from reStructuredText in those blocks that use it for inline formatting. 

See 
https://docutils.sourceforge.io/docs/user/rst/quickref.html 
or 
https://github.com/ralsina/rst-cheatsheet 
for a quick lookup of what's in the format

**Comment by tremblap** [comment]
thanks.

**Comment by tremblap** [comment]
ok I love it. I was hoping for auto indent if I enter .. parameter:: in [enter] but hey. Looking at the cheat sheets I can see how much clearer that would make it all so I vote for it.

**Comment by weefuzzy** [comment]
I don't think it would have a way of knowing that you want child content in a directive, because it's not mandatory. Such are the challenges with formats that use whitespace for semantics rather than control characters: what you win in readability and brevity, you lose a bit in automatic anticipation. 

Anyway, glad everyone likes this idea. It will come after #24 is done.

**Comment by jamesb93** [comment]
🔥 
vscode naturally supports it well too so I feel well supported in that regard. woohoo
